### <span style='color:Red'>Linear Regression </span>
###### mean squared error(MSE)

In [5]:
from sklearn.model_selection import train_test_split 
from sklearn.metrics import mean_squared_error
import numpy as np


##### sample data

In [6]:
x=np.array([[1],[2],[3],[4],[5]])
print(x)
y=np.array([2,4,5,4,5])
print(y)

[[1]
 [2]
 [3]
 [4]
 [5]]
[2 4 5 4 5]


In [7]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=42)


##### Mean Squared Error (MSE)
### MSE=1/n∑(yi−y^i)2

In [11]:
from sklearn.linear_model import LinearRegression
lr=LinearRegression()
lr.fit(x_train,y_train)
y_pred=lr.predict(x_test)
mse=mean_squared_error(y_test,y_pred)
print("y_test:",y_test)
print("y_pred:",y_pred)
print("Mean squared Error on test data:",mse)

y_test: [4]
y_pred: [3.14285714]
Mean squared Error on test data: 0.7346938775510206


#### Predicted = 3.14
#### Actual = 4

### Lasso Regression
##### Lasso regression is Linear Regression with L1 regularization:
#### J(β)=∑(yi−y^i)2+λ∑∣βj∣

#### Lasso stands for Least Absolute Shrinkage and Selection Operator. It is a type of linear regression that includes L1 regularization.
### loss = ∑(yi- yi^)^2 + α ∑|Wj|

### Alpha (α)
##### Alpha is the regularization parameter in Lasso.
##### Controls the strength of the penalty on coefficients:

##### α = 0 → Lasso becomes ordinary linear regression (no penalty).
##### α large → Stronger penalty → more coefficients shrink to zero.

import numpy as np
from sklearn.linear_model import Lasso
from sklearn.metrics import mean_squared_error,r2_score
x=np.array([[1],[2],[3],[4],[5]])
y=np.array([2,3,5,7,11])
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=42)


### Lasso Regression

In [18]:
lasso_model=Lasso(alpha=0.1)
lasso_model.fit(x_train,y_train)
y_pred_lasso=lasso_model.predict(x_test)
print("\n Lasso Regression:")
print("Predictions:", y_pred_lasso)
print("MSE:",mean_squared_error(y_test,y_pred_lasso))
print("R2 score:",r2_score(y_test,y_pred_lasso))


 Lasso Regression:
Predictions: [3.62857143]
MSE: 0.39510204081632594
R2 score: nan


E:\anaconda for jupyter\Lib\site-packages\sklearn\metrics\_regression.py:1266: UndefinedMetricWarning: R^2 score is not well-defined with less than two samples.
  warnings.warn(msg, UndefinedMetricWarning)


#### R² score tells you how well your model fits the data.
#### R² = 1 → perfect prediction, 0 → predicts just the mean, negative → worse than mean.
#### NaN appears if your test set has only 1 sample, because variance of a single number = 0.

## Ridge Regression
##### Ridge Regression is a type of linear regression that includes L2 regularization. It’s used to prevent overfitting when your features are correlated or when you have many features.

In [21]:
from sklearn.linear_model import Ridge
ridge_model=Ridge(alpha=1.0)
ridge_model.fit(x_train,y_train)
y_pred_ridge=ridge_model.predict(x_test)
print("Ridge Regression")
print("Predictions: " ,y_pred_ridge)
print("MSE:", mean_squared_error(y_test,y_pred_ridge))
print("R2 Score:",r2_score(y_test,y_pred_ridge))

Ridge Regression
Predictions:  [3.84615385]
MSE: 0.7159763313609476
R2 Score: nan


E:\anaconda for jupyter\Lib\site-packages\sklearn\metrics\_regression.py:1266: UndefinedMetricWarning: R^2 score is not well-defined with less than two samples.
  warnings.warn(msg, UndefinedMetricWarning)
